File: day03/python/main.py
INPUT_FILE = "../input.txt"

def to_code(char):
    code = ord(char)
    if code > 96:
        return code - 96
    else:
        return code - 38

part1 = 0
part2 = 0

with open(INPUT_FILE, 'r') as file:
    buffers = [
        set(), set()
    ]
    for i, line in enumerate(file):
        x = i % 3
        line = line.strip()
        half = int(len(line)/2)
        a = set([to_code(x) for x in line[:half]])
        b = set([to_code(x) for x in line[half:]])
        part1 += sum(a.intersection(b))
        if x == 2:
            temp = buffers[0].intersection(buffers[1]).intersection(a | b)
            part2 += sum(temp)
        else:
            buffers[x] = a | b

print('Part #1:', part1)
print('Part #2:', part2)

File: day03/rust/src/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};
use std::collections::HashSet;


const INPUT_FILE: &'static str = "../input.txt";

type Result<T> = ::std::result::Result<T, Box<dyn ::std::error::Error>>;
 
fn main() -> Result<()> {
    let file = File::open(INPUT_FILE)?;
    let reader = BufReader::new(file);
    let lines = reader.lines().flatten();

    let mut part1: u32 = 0;
    let mut part2: u32 = 0;
    let mut b: HashSet<u8> = HashSet::new();
    let mut buffers: [HashSet<u8>; 3] = [
        HashSet::new(), HashSet::new(), HashSet::new(),
    ];

    for (i, line) in lines.enumerate() {
        let n = i % 3;
        let length = line.len();
        let a: &mut HashSet<u8> = &mut buffers[n];
        a.clear();
        b.clear();
        a.extend(line.chars().map(|x| to_code(&x)).take(length / 2));
        b.extend(line.chars().map(|x| to_code(&x)).skip(length / 2));
        part1 += a.intersection(&b).sum::<u8>() as u32;
        a.extend(&b);

        if n == 2 {
            let temp: HashSet<u8> = buffers[0].intersection(&buffers[1]).cloned().collect();
            part2 += temp.intersection(&buffers[2]).sum::<u8>() as u32;
        }
    }

    println!("Part #1: {:?}", part1);
    println!("Part #2: {:?}", part2);

    Ok(())
}

fn to_code(c: &char) -> u8 {
    match *c as u8 {
        c if c > 96 => { c - 96 }
        c => { c - 38 }
    }
}
